src/test/java/com/google/sps/CommentServiceTest.java

```diff
@@ -0,0 +1,109 @@
+package com.google.sps;
+
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.services.youtube.YouTube;
+import com.google.api.services.youtube.model.CommentThread;
+import com.google.api.services.youtube.model.CommentThreadListResponse;
+import com.google.sps.servlets.CommentService;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Answers;
+import org.mockito.InjectMocks;
+import org.mockito.Matchers;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import static org.mockito.ArgumentMatchers.*;
+import static org.mockito.Mockito.*;
+
+@RunWith(MockitoJUnitRunner.class)
+
+public final class CommentServiceTest {
+
+  String VIDEO_ID = ""; // Arbitrary video ID, never used.
+
+  @Mock CommentThreadListResponse responseMock;
+  @Mock(answer = Answers.RETURNS_DEEP_STUBS) YouTube youtubeServiceMock;
+  @Mock(answer = Answers.RETURNS_DEEP_STUBS) YouTube.CommentThreads.List requestMock;
+  @Mock(answer = Answers.RETURNS_DEEP_STUBS) CommentThread commentThreadMock;
+  @Mock(answer = Answers.RETURNS_DEEP_STUBS) GoogleJsonResponseException exceptionMock;
+
+  @InjectMocks CommentService commentService;
+
+  @Before
+  public void setUp() throws IOException {
+    when(youtubeServiceMock.commentThreads().list(any())).thenReturn(requestMock);
+  }
+
+  /**
+   * A list of Strings should be returned for any public/unlisted video that has comments. 
+   */
+  @Test
+  public void videoReturnsCommentsList() throws GeneralSecurityException, IOException {
+    List<CommentThread> threadList =
+        new ArrayList<>(Arrays.asList(commentThreadMock, commentThreadMock));
+    
+    when(requestMock.setKey(anyString())
+        .setMaxResults(anyLong())
+        .setOrder("relevance")
+        .setTextFormat("plainText")
+        .setVideoId(anyString())
+        .execute())
+        .thenReturn(responseMock);
+    when(responseMock.getItems()).thenReturn(threadList);
+    when(commentThreadMock.getSnippet().getTopLevelComment().getSnippet().getTextDisplay())
+        .thenReturn("a", "b"); // Returns "a" first time, "b" second time.
```

Comment:
Hmm I'm not very familiar with Java mocking but I feel like this might be too many mocks. You want to minimize mocks only to when you call something outside this service. In this PR, that would be building the youtube api service and calling the list comment threads method. You are doing these two with the `when` call in the setup method and the first one here. The remaining `when` calls here are used to mock the response you already mocked. What you want to test is how your code handles the response from the API. Mocking anything after the response, means you might be faking too much. 

Is there a way to create an actual an actual CommentThread List object and then force to be returned in the first `when` in this state such that when it calls getItems, getSnippet, getTopLevelComment, getSnippet, and getTextDisplay it all works naturally?

Creating this object to be returned might involve a lot of setting up, so I recommend creating a helper method where the arguments are the relevant data needed for the test. In this case, the relevant data would be `"a"` and `"b"`.

Copied this internal doc for the pod. Mock Object Best Practices https://drive.google.com/file/d/1W9iZyu5kYOTUjO6_WHTNqUy4ie6FmHyc/view?usp=sharing

If you want to work on this together, feel free to set up some time on my calendar and you can screen share what you are doing. Meeting like this reduces the friction between PR reviews. I recommend you first attempt it yourself so there isn't a lot of waiting time during our meeting.

Reply:
Thanks for the offer! I think I figured out what you meant, your description was super helpful :)